pxtcompiler/emitter/transpile.ts

```diff
@@ -0,0 +1,115 @@
+// TODO: enable reference so we don't need to use: (pxt as any).py
+//      the issue is that this creates a circular dependency. This
+//      is easily handled if we used proper TS modules.
+//// <reference path="../../built/pxtpy.d.ts"/>
+
+// This code centralizes transpiling between typescript and python
+// so that we can cache results for better performance and user
+// experience (since compile -> decompile round-trips are lossy)
+namespace ts.pxtc.transpile {
+    export interface TranspileResult {
+        diagnostics: pxtc.KsDiagnostic[],
+        success: boolean,
+        outfiles: { [key: string]: string }
+    }
+
+    const mainName = (l: Lang) => `main.${l}`
+
+    export type Lang = "py" | "blocks" | "ts"
+    export interface LangEquivSet {
+        comparable: { [key in Lang]: number },
+        code: { [key in Lang]: string },
+    }
+    // a circular buffer of size MAX_CODE_EQUIVS that stores
+    // sets of equivalent code files so that when we translate
+    // from ts -> py -> ts we get back the same ts.
+    let codeEquivalences: LangEquivSet[] = []
+    const MAX_CODE_EQUIVS = 10
+
+    function toComparable(code: string): number {
+        // Ignore whitespace
+        code = code.replace(/\s/g, "")
+        return U.codalHash16(code)
+    }
+    function tryGetLanguageEquivalence(lang: Lang, txt: string): LangEquivSet | undefined {
+        let txtComp = toComparable(txt)
+        for (let eq of codeEquivalences) {
+            if (eq.comparable[lang] === txtComp) {
+                return eq
+            }
+        }
+        return undefined
+    }
+    function setLanguageEquivalence(lang1: Lang, lang1Txt: string, lang2: Lang, lang2Txt: string) {
```

Comment:
I think a clearer name would be cacheLanguageEquivalence or even just cacheTranspile. Took me a bit to realize what was going on in this file.